File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/SemanticPlan.java
package com.foundgine.core.semantic.planning;

import com.foundgine.core.semantic.authorization.SemanticAuthorizationEvidence;
import com.foundgine.core.semantic.SemanticContractSnapshot;
import java.util.*;

/** Canonical provider-neutral semantic planning artifact. */
public record SemanticPlan(SemanticPlanNode root, List<String> requiredSecurityInvariants,
		SemanticPlanAuthorizationBinding authorizationBinding) {
	public SemanticPlan {
		Objects.requireNonNull(root);
		requiredSecurityInvariants = requiredSecurityInvariants == null ? List.of()
				: List.copyOf(requiredSecurityInvariants);
	}

	public SemanticPlan(SemanticPlanNode root) {
		this(root, List.of(), null);
	}

	public List<String> effectiveSecurityInvariants() {
		return requiredSecurityInvariants;
	}
}

File: src/csharp/Foundgine.Core/Semantic/Planning/SemanticPlan.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Canonical semantic planning artifact produced from an authorized Semantic IR.
/// It describes the provider-neutral execution strategy without representing
/// physical provider work.
/// </summary>
public sealed record SemanticPlan(
    SemanticPlanNode Root,
    IReadOnlyList<string>? RequiredSecurityInvariants = null,
    SemanticPlanAuthorizationBinding? AuthorizationBinding = null)
{
    public IReadOnlyList<string> EffectiveSecurityInvariants =>
        RequiredSecurityInvariants ?? [];
}

/// <summary>
/// One node in the canonical semantic planning artifact.
/// </summary>
public sealed record SemanticPlanNode(
    int Id,
    ExecutionOperation Operation,
    EntityId EntityId,
    IReadOnlyList<FieldId> Fields,
    RelationshipId? ViaRelationship,
    ConnectionId? ViaConnection,
    IReadOnlyList<SemanticPlanNode> Children,
    Foundgine.Core.Semantic.Query.SemanticQueryOptions? QueryOptions = null,
    AuthorizationPredicate? Authorization = null,
    Foundgine.Core.Semantic.RelationshipCardinality? RelationshipCardinality = null,
    RelationshipTraversalMode TraversalMode = RelationshipTraversalMode.Default,
    int TraversalOrder = -1,
    AggregateExecutionStrategy AggregateExecutionStrategy = AggregateExecutionStrategy.Default)
{
    public SemanticPlanNode(
        int id,
        ExecutionOperation operation,
        EntityId entityId,
        IReadOnlyList<FieldId> fields,
        RelationshipId? viaRelationship,
        IReadOnlyList<SemanticPlanNode> children)
        : this(id, operation, entityId, fields, viaRelationship, null, children, null)
    {
    }
}
